File: backend/src/main/java/com/locationservice/service/ActivityService.java
package com.locationservice.service;

import com.locationservice.dto.ActivityDto;
import com.locationservice.entity.Activity;
import com.locationservice.entity.DeviceStatus;
import com.locationservice.entity.Location;
import com.locationservice.entity.User;
import com.locationservice.repository.ActivityRepository;
import com.locationservice.repository.DeviceStatusRepository;
import com.locationservice.repository.LocationRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class ActivityService {

    private final ActivityRepository activityRepository;
    private final LocationRepository locationRepository;
    private final DeviceStatusRepository deviceStatusRepository;

    private static final DateTimeFormatter ISO_FORMATTER = DateTimeFormatter.ISO_LOCAL_DATE_TIME;

    public ActivityService(
            ActivityRepository activityRepository,
            LocationRepository locationRepository,
            DeviceStatusRepository deviceStatusRepository) {
        this.activityRepository = activityRepository;
        this.locationRepository = locationRepository;
        this.deviceStatusRepository = deviceStatusRepository;
    }

    @Transactional
    public List<ActivityDto> getActivities(User currentUser, String type) {
        List<Activity> activities;

        if (type != null && !type.isBlank()) {
            String filterType = normalizeType(type);
            activities = activityRepository.findByUserAndTypeOrderByCreatedAtDesc(currentUser, filterType);
        } else {
            activities = activityRepository.findByUserOrderByCreatedAtDesc(currentUser);
        }

        // Seed initial activity history records if user has no activities in PostgreSQL
        if ((activities == null || activities.isEmpty()) && (type == null || type.isBlank())) {
            activities = seedInitialActivities(currentUser);
        }

        if (activities == null) {
            activities = new ArrayList<>();
        }

        Optional<Location> latestLoc = locationRepository.findTopByUserOrderByUpdatedAtDesc(currentUser);
        Optional<DeviceStatus> latestDevice = deviceStatusRepository.findByUser(currentUser);

        Double currentLat = latestLoc.map(Location::getLatitude).orElse(28.6139);
        Double currentLng = latestLoc.map(Location::getLongitude).orElse(77.2090);
        String currentAddress = latestLoc.map(Location::getAddress).orElse(String.format("%.4f, %.4f", currentLat, currentLng));
        Integer batteryLevel = latestDevice.map(DeviceStatus::getBatteryLevel).orElse(95);

        return activities.stream()
                .map(act -> mapToDto(act, currentUser, currentLat, currentLng, currentAddress, batteryLevel))
                .collect(Collectors.toList());
    }

    private ActivityDto mapToDto(Activity activity, User currentUser, Double currentLat, Double currentLng, String currentAddress, Integer batteryLevel) {
        Double lat = activity.getLatitude() != null ? activity.getLatitude() : currentLat;
        Double lng = activity.getLongitude() != null ? activity.getLongitude() : currentLng;
        LocalDateTime createdAt = activity.getCreatedAt() != null ? activity.getCreatedAt() : LocalDateTime.now();

        return new ActivityDto(
                activity.getId(),
                activity.getTitle() != null ? activity.getTitle() : "Telemetry Log",
                activity.getDetails() != null ? activity.getDetails() : "Activity registered.",
                activity.getType() != null ? activity.getType() : "GPS_SYNC",
                lat,
                lng,
                currentAddress,
                currentUser.getName() != null ? currentUser.getName() : "System Node",
                "Android",
                batteryLevel,
                createdAt.format(ISO_FORMATTER)
        );
    }

    private List<Activity> seedInitialActivities(User currentUser) {
        LocalDateTime now = LocalDateTime.now();
        List<Activity> seeded = new ArrayList<>();

        seeded.add(activityRepository.save(new Activity(
                currentUser,
                "GPS Telemetry Synced",
                "Real-time GPS coordinates synchronized with central server.",
                "GPS_SYNC",
                28.6139,
                77.2090,
                now.minusMinutes(5)
        )));

        seeded.add(activityRepository.save(new Activity(
                currentUser,
                "PTT Voice Intercom Call",
                "Push-to-Talk voice channel session established (0:14s duration).",
                "PTT_VOICE",
                28.6145,
                77.2095,
                now.minusMinutes(25)
        )));

        seeded.add(activityRepository.save(new Activity(
                currentUser,
                "Camera Telemetry HUD Snapshot",
                "Live camera optical snapshot captured and location tagged.",
                "CAMERA_TELEMETRY",
                28.6150,
                77.2100,
                now.minusHours(1).minusMinutes(10)
        )));

        seeded.add(activityRepository.save(new Activity(
                currentUser,
                "Geofence Security Perimeter Check",
                "Entered designated safe zone boundary at Checkpoint Alpha.",
                "GEOFENCE",
                28.6155,
                77.2105,
                now.minusHours(2)
        )));

        return activityRepository.findByUserOrderByCreatedAtDesc(currentUser);
    }

    private String normalizeType(String inputType) {
        String upper = inputType.trim().toUpperCase();
        if ("LOCATION".equals(upper) || "LOCATION_UPDATE".equals(upper)) return "GPS_SYNC";
        if ("PTT".equals(upper) || "PTT_CALL".equals(upper)) return "PTT_VOICE";
        if ("SOS_ALERT".equals(upper)) return "SOS";
        if ("SNAPSHOT".equals(upper) || "CAMERA_SNAPSHOT".equals(upper)) return "CAMERA_TELEMETRY";
        if ("GEOFENCE_ALERT".equals(upper)) return "GEOFENCE";
        return upper;
    }
}

File: backend/src/services/activity.service.ts
import { prisma } from '../config/database.js';
import { AuthenticatedUser } from '../middleware/auth.middleware.js';

export class ActivityService {
  public static async getActivities(currentUser: AuthenticatedUser, type?: string) {
    let activities;

    if (type && type.trim().length > 0) {
      const normalized = this.normalizeType(type);
      activities = await prisma.activity.findMany({
        where: { userId: currentUser.id, type: normalized },
        orderBy: { createdAt: 'desc' },
      });
    } else {
      activities = await prisma.activity.findMany({
        where: { userId: currentUser.id },
        orderBy: { createdAt: 'desc' },
      });
    }

    if (activities.length === 0 && (!type || type.trim().length === 0)) {
      activities = await this.seedInitialActivities(currentUser);
    }

    const latestLoc = await prisma.location.findFirst({
      where: { userId: currentUser.id },
      orderBy: { updatedAt: 'desc' },
    });

    const latestDevice = await prisma.deviceStatus.findUnique({
      where: { userId: currentUser.id },
    });

    const currentLat = latestLoc ? latestLoc.latitude : 28.6139;
    const currentLng = latestLoc ? latestLoc.longitude : 77.2090;
    const currentAddress = latestLoc?.address || `${currentLat.toFixed(4)}, ${currentLng.toFixed(4)}`;
    const batteryLevel = latestDevice ? latestDevice.batteryLevel : 95;

    return activities.map((act) => {
      const lat = act.latitude !== null && act.latitude !== undefined ? act.latitude : currentLat;
      const lng = act.longitude !== null && act.longitude !== undefined ? act.longitude : currentLng;
      const createdAt = act.createdAt ? act.createdAt.toISOString().slice(0, 19) : new Date().toISOString().slice(0, 19);

      return {
        id: Number(act.id),
        title: act.title || 'Telemetry Log',
        details: act.details || 'Activity registered.',
        type: act.type || 'GPS_SYNC',
        latitude: lat,
        longitude: lng,
        address: currentAddress,
        userName: currentUser.name || 'System Node',
        deviceModel: 'Android',
        batteryLevel,
        createdAt,
      };
    });
  }

  private static async seedInitialActivities(currentUser: AuthenticatedUser) {
    const now = new Date();

    const seeds = [
      {
        userId: currentUser.id,
        title: 'GPS Telemetry Synced',
        details: 'Real-time GPS coordinates synchronized with central server.',
        type: 'GPS_SYNC',
        latitude: 28.6139,
        longitude: 77.2090,
        createdAt: new Date(now.getTime() - 5 * 60 * 1000),
      },
      {
        userId: currentUser.id,
        title: 'PTT Voice Intercom Call',
        details: 'Push-to-Talk voice channel session established (0:14s duration).',
        type: 'PTT_VOICE',
        latitude: 28.6145,
        longitude: 77.2095,
        createdAt: new Date(now.getTime() - 25 * 60 * 1000),
      },
      {
        userId: currentUser.id,
        title: 'Camera Telemetry HUD Snapshot',
        details: 'Live camera optical snapshot captured and location tagged.',
        type: 'CAMERA_TELEMETRY',
        latitude: 28.6150,
        longitude: 77.2100,
        createdAt: new Date(now.getTime() - (70 * 60 * 1000)),
      },
      {
        userId: currentUser.id,
        title: 'Geofence Security Perimeter Check',
        details: 'Entered designated safe zone boundary at Checkpoint Alpha.',
        type: 'GEOFENCE',
        latitude: 28.6155,
        longitude: 77.2105,
        createdAt: new Date(now.getTime() - (120 * 60 * 1000)),
      },
    ];

    for (const seed of seeds) {
      await prisma.activity.create({ data: seed });
    }

    return prisma.activity.findMany({
      where: { userId: currentUser.id },
      orderBy: { createdAt: 'desc' },
    });
  }

  private static normalizeType(inputType: string): string {
    const upper = inputType.trim().toUpperCase();
    if (upper === 'LOCATION' || upper === 'LOCATION_UPDATE') return 'GPS_SYNC';
    if (upper === 'PTT' || upper === 'PTT_CALL') return 'PTT_VOICE';
    if (upper === 'SOS_ALERT') return 'SOS';
    if (upper === 'SNAPSHOT' || upper === 'CAMERA_SNAPSHOT') return 'CAMERA_TELEMETRY';
    if (upper === 'GEOFENCE_ALERT') return 'GEOFENCE';
    return upper;
  }
}
